pkg/kv/kvserver/kvstorage/init.go

```diff
@@ -291,103 +293,197 @@ func IterateRangeDescriptorsFromDisk(
 	return err
 }
 
-type EngineReplicas struct {
-	Uninitialized map[storage.FullReplicaID]struct{}
-	Initialized   map[storage.FullReplicaID]*roachpb.RangeDescriptor
+// A Replica references a CockroachDB Replica. The data in this struct does not
+// represent the data of the Replica but is sufficient to access all of its
+// contents via additional calls to the storage engine.
+type Replica struct {
+	RangeID   roachpb.RangeID
+	ReplicaID roachpb.ReplicaID
+	Desc      *roachpb.RangeDescriptor // nil for uninitialized Replica
+
+	hardState raftpb.HardState // internal to kvstorage, see migration in LoadAndReconcileReplicas
 }
 
-// loadFullReplicaIDsFromDisk discovers all Replicas on this Store.
-// There will only be one entry in the map for a given RangeID.
-//
-// TODO(sep-raft-log): the reader here is for the log engine.
-func loadFullReplicaIDsFromDisk(
-	ctx context.Context, reader storage.Reader,
-) (map[storage.FullReplicaID]struct{}, error) {
-	m := map[storage.FullReplicaID]struct{}{}
-	var msg roachpb.RaftReplicaID
-	if err := IterateIDPrefixKeys(ctx, reader, func(rangeID roachpb.RangeID) roachpb.Key {
-		return keys.RaftReplicaIDKey(rangeID)
-	}, &msg, func(rangeID roachpb.RangeID) error {
-		m[storage.FullReplicaID{RangeID: rangeID, ReplicaID: msg.ReplicaID}] = struct{}{}
-		return nil
-	}); err != nil {
-		return nil, err
+// ID returns the FullReplicaID.
+func (r Replica) ID() storage.FullReplicaID {
+	return storage.FullReplicaID{
+		RangeID:   r.RangeID,
+		ReplicaID: r.ReplicaID,
+	}
+}
+
+// A ReplicaMap organizes a set of Replicas with unique RangeIDs.
+type ReplicaMap map[roachpb.RangeID]Replica
+
+// Sorted returns the contents of the map as a RangeID-sorted slice.
+func (m ReplicaMap) Sorted() []Replica {
+	sl := make([]Replica, 0, len(m))
+	for _, repl := range m {
+		sl = append(sl, repl)
 	}
+	sort.Slice(sl, func(i, j int) bool {
+		return sl[i].RangeID < sl[j].RangeID
+	})
+	return sl
+}
+
+func (m ReplicaMap) getOrMake(rangeID roachpb.RangeID) Replica {
+	ent := m[rangeID]
+	ent.RangeID = rangeID
+	return ent
+}
+
+func (m ReplicaMap) setReplicaID(rangeID roachpb.RangeID, replicaID roachpb.ReplicaID) Replica {
+	ent := m.getOrMake(rangeID)
+	ent.ReplicaID = replicaID
+	m[rangeID] = ent
+	return ent
```

Comment:
nil: dunno, this is a bit weird, can we avoid this API?

I think (idea 1):
- The user of `LoadAndReconcileReplicas` doesn't need the map, so it can return `[]Replica` directly.
- Since we're sorting anyway (for determinism), let it return a sorted `[]Replica`.
- Then, the tests don't need to `Sorted()` the map again.
- Also, `ReplicaMap` and all its methods (incl. `Sorted`) can be unexported.
- Inside `LoadAndReconcileReplicas` in the place where we start overriding things in the map, we should just convert to sorted slice, and modify slice elements directly.

Idea 2: If you really want to expose the map-ness to the outside (e.g. the user of this needs to search by RangeID, which though it currently doesn't), I could also imagine a data structure like this:

```golang
type ReplicaMap struct {
	index map[roachpb.RangeID]int
	repls []Replica
}

func (m *ReplicaMap) getOrMake(rangeID roachpb.RangeID) int { // or *Replica
	if index, found := m[rangeID]; found {
		return index
	}
	m.repls = append(m.repls, Replica{RangeID: rangeID})
	return len(m.repls) - 1
}

func (m *ReplicaMap) Sort() { // or stateless Sorted() like right now
	// sort m.repls
	// reindex m.index using m[].RangeID as key
}

// then all the accessors are not needed, can just `getOrMake` and write
// to the relevant returned *Replica (or repls[index]) directly, like this:
// 
m.getOrMake(rangeID).hardState = hs
// or
m.repls[m.getOrMake(rangeID)].hardState = hs
```

Some combination of idea 1 and 2 would be nice.

Reply:
Went with 1), the yak has been shaved 👑 
I like the result better.